src/Temporalio/Worker/ActivityWorker.cs

```diff
@@ -0,0 +1,703 @@
+                // shutdown which could never run until after this (and the primary execute) are
+                // done. So we don't have to worry about the dictionary having an activity without
+                // a task even though we're late-binding it here.
+                var act = new RunningActivity(context, cancelTokenSource);
+                runningActivities[tsk.TaskToken] = act;
+#pragma warning disable CA2008 // We don't have to pass a scheduler, factory already implies one
+                act.Task = worker.Options.ActivityTaskFactory.StartNew(
+                    () => ExecuteActivityAsync(act, tsk)).Unwrap();
+#pragma warning restore CA2008
+            }
+        }
+
+        private async Task ExecuteActivityAsync(
+            RunningActivity act,
+            Bridge.Api.ActivityTask.ActivityTask tsk)
+        {
+            // Try to build a completion, but if it fails then manually build one with the failure
+            Bridge.Api.ActivityTaskCompletion completion;
+            try
+            {
+                completion = await ExecuteActivityInternalAsync(act, tsk).ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                completion = new()
+                {
+                    Result = new()
+                    {
+                        Failed = new()
+                        {
+                            Failure_ = new() { Message = $"Failed building completion: {e}" },
+                        },
+                    },
+                };
+            }
+
+            // Try to finish heartbeats and send completion
+            try
+            {
+                // We have to wait on any outstanding heartbeats to finish. This will not throw. We
+                // accept that in a rare scenario, this heartbeat can fail to encode but it is too
+                // late to cancel the activity. Like other SDKs, we currently drop this heartbeat.
+                await act.FinishHeartbeatsAsync().ConfigureAwait(false);
+
+                // Complete the task
+                act.Context.Logger.LogTrace("Sending activity completion: {Completion}", completion);
+                await worker.BridgeWorker.CompleteActivityTaskAsync(
+                    completion).ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                act.Context.Logger.LogError(
+                    e,
+                    "Failed completing activity task with completion {Completion}",
+                    completion);
+            }
+            finally
+            {
+                // Remove from running activities
+                runningActivities.TryRemove(tsk.TaskToken, out _);
+            }
+        }
+
+        private async Task<Bridge.Api.ActivityTaskCompletion> ExecuteActivityInternalAsync(
+            RunningActivity act,
+            Bridge.Api.ActivityTask.ActivityTask tsk)
+        {
+            act.Context.Logger.LogDebug(
```

Comment:
Doesn't this run in the activity info logger scope? If so, you should have the activity type in the scope.

Reply:
Showing scope in .NET is opt-in and many loggers don't even support it. That's why for non-internal-error type logs of an activity, I also qualify the activity a bit in the log string.

Reply:
Got it